pkg/app/validate_app.go

```diff
@@ -118,7 +126,55 @@ func (v *Validator) Validate(request *admissionv1.AdmissionRequest) (bool, error
 		return false, microerror.Mask(err)
 	}
 
+	err = v.emitEvents(ctx, request, app)
```

Comment:
Only emit events when it's about to be allowed.

Reply:
Feels like this point is important. Can you add comments for it?